src/re/multi_grid/kernel.py

```diff
@@ -1,282 +1,276 @@
+        )
+        super().__init__(grid=grid)
+
+    @property
+    def covariance_outer(self):
+        return self._covariance_outer
+
+    @property
+    def window_size(self):
+        return self._window_size
+
+    def get_output_input_indices(self, index, level):
+        if level is None:
+            grid_at_lvl = self.grid.at(0)
             pixel_indices = np.mgrid[tuple(slice(0, sz) for sz in grid_at_lvl.shape)]
             assert pixel_indices.shape[0] == grid_at_lvl.ndim
             return (pixel_indices, 0), (
                 (pixel_indices.reshape(grid_at_lvl.ndim, -1), 0),
             )
-        if (level >= self._grid.depth) or (level < -1):
-            mg = f"Level {level} out of bounds for grid deph {self._grid.depth}"
-            raise ValueError(mg)
-        atlevel = self._grid.at(level)
-        assert index.shape[0] == atlevel.ndim
-        gc = atlevel.neighborhood(index, self._window_size).reshape(index.shape + (-1,))
-        gout = self._grid.at(level).children(index)
+        g = self.grid.at(level)
+        assert index.shape[0] == g.ndim
+        gc = g.neighborhood(index, self.window_size[level]).reshape(index.shape + (-1,))
+        gout = g.children(index)
         gf = gout.reshape(index.shape + (-1,))
         return (gout, level + 1), ((gc, level), (gf, level + 1))
 
-    def get_kernel(self, index, level):
-        if level == -1:
-            _, ((ids, _),) = self.get_indices(index, -1)
-            gc = self._grid.at(0).index2coord(ids)
-            assert gc.ndim == 2
-            cov = self._covariance(gc, gc)
+    def get_matrices(self, index, level):
+        from ..refine.util import sqrtm
+
+        if level is None:
+            _, ((ids, _),) = self.get_output_input_indices(index, None)
+            gc = self.grid.at(0).index2coord(ids)
+            cov = self.covariance_outer(gc, gc)
             assert cov.shape == (gc.shape[1],) * 2
-            from ..refine.util import projection_MatrixSq
+            return (sqrtm(cov),)
 
-            return (projection_MatrixSq(cov),)
-        _, ((idc, _), (idf, _)) = self.get_indices(index, level)
+        _, ((idc, _), (idf, _)) = self.get_output_input_indices(index, level)
 
-        def _get_kernel(gc, gf):
+        def get_mat(gc, gf):
             gc = self.grid.at(level).index2coord(gc)
             gf = self.grid.at(level + 1).index2coord(gf)
             assert gc.shape[0] == gf.shape[0]
             assert gc.ndim == gf.ndim == 2
             coord = jnp.concatenate((gc, gf), axis=-1)
-            cov = self._covariance(coord, coord)
+            cov = self.covariance_outer(coord, coord)
             return refinement_matrices(cov, gf.shape[1])
 
-        f = _get_kernel
+        f = get_mat
         for _ in range(index.ndim - 1):
             f = vmap(f, in_axes=(1, 1))
         return f(idc, idf)
 
 
-class FixedKernelFunctionBatch(KernelBase):
+class FixedKernelFunctionBatch(Kernel):
```

Comment:
I believe that all Kernels from here on are also still under development and not fully functional. I suggest to not merge them in this PR also?

Reply:
My intention was to delete everything that is not used once we are happy with the user facing changes. As these kernels are almost certainly not part of any user facing changes, I am happy to delete them right away.